### PR TITLE
Introduce ClientSession configuration section that provides ways to control idleTime and lifeTime of client connections.

### PR DESCRIPTION
**Pull request by nikolay-pshenichny**
This PR introduces a new client configuration section that allows to control `idleTime` and `lifeTime` of the client connections.

Fixes https://github.com/linkerd/linkerd/issues/1890

### CODE REVIEW

**Comment by nikolay-pshenichny** [comment]
@adleong , done.

File: linkerd/docs/routers.md [hunk 1]
```diff
@@ -269,5 +273,14 @@ Key | Default Value | Description
 --- | ------------- | -----------
 minSize | `0` | The minimum number of connections to maintain to each host.
 maxSize | Int.MaxValue | The maximum number of connections to maintain to each host.
-idleTimeMs | forever | The amount of idle time for which a connection is cached in milliseconds.
+idleTimeMs | forever | The amount of idle time for which a connection is cached in milliseconds. Only applied to connections that number greater than minSize, but fewer than maxSize.
```
Comment (dadjeibaah) [inline linkerd/docs/routers.md]: I am trying to understand this comment. How do `HostConnectionPool`'s idleTimeMs differ from `ClientSession`'s idleTimeMs? Do they effectively do the same thing? If they do, then we may no longer need to have this setting in `HostConnectionPool`.

File: linkerd/protocol/http/src/e2e/scala/io/buoyant/linkerd/protocol/HttpEndToEndTest.scala [hunk 2]
```diff
@@ -693,4 +697,152 @@ class HttpEndToEndTest
...
+
+    } finally {
+      await(client.close())
+      await(fox.server.close())
+      await(server.close())
+      await(router.close())
+    }
+
+
+  }
```
Comment (dadjeibaah) [inline linkerd/protocol/http/src/e2e/scala/io/buoyant/linkerd/protocol/HttpEndToEndTest.scala]: I may be wrong, but maybe we could add a test that exercises both `hostConnectionPool` and `clientSession` configs. I am still unclear of how the two different kinds of configurations would work together if that ever is the case.

File: linkerd/docs/routers.md (hunk 1 above)
Comment (adleong) [reply linkerd/docs/routers.md]: They are similar but slightly different.  This connection timeout is only applied to connections between the minSize and maxSize whereas the clientSessions timeouts apply to all connections.  The connectionPool timeout can be thought of as the amount of time "extra" connections are kept around.

Comment (adleong) [reply linkerd/docs/routers.md]: Although, in the case where minSize=0 (which is the default) then these two timeouts do exactly the same thing.

Comment (dadjeibaah) [reply linkerd/docs/routers.md]: Ahh I see, thanks for the clarification!

**Comment by adleong** [comment]
@nikolay-pshenichny Looks like there's a merge conflict with the latest master.  Would you mind resolving that and then we should be good to merge.

**Comment by nikolay-pshenichny** [comment]
@adleong , I will look into that a little later today.

File: linkerd/protocol/http/src/e2e/scala/io/buoyant/linkerd/protocol/HttpEndToEndTest.scala (hunk 2 above)
Comment (nikolay-pshenichny) [reply linkerd/protocol/http/src/e2e/scala/io/buoyant/linkerd/protocol/HttpEndToEndTest.scala]: @dadjeibaah , I can look into adding more tests for the scenarios when both configurations are used. But I will be able to do that only starting Tuesday (24th) next week.

**Comment by adleong** [comment]
:star: Thank you so much for tracking this down and fixing it, @nikolay-pshenichny!

**Pull request merged**